File: tools/external/python-escpos/capture.py
#!/usr/bin/env python3

"""Capture the byte streams of the python-escpos examples, section 16.

python-escpos has a ``Dummy`` printer that collects the bytes a script writes
without a device, so an example runs without hardware.  This script fetches the
examples of its repository at a pinned commit, runs each of them against a
``Dummy`` and writes what came out to
``test/fixtures/external/python-escpos/<example>.bin``.

The examples that construct a ``Usb`` printer are run with a ``Dummy`` of the
same profile instead; nothing else about them is changed.  The examples that
need a network service are not captured, see SKIPPED below.

Run it in a throwaway virtual environment with the library installed from the
same commit as the examples, so that the two match; the provenance records the
version that reports:

    python3 -m venv build/external/python-escpos/venv
    build/external/python-escpos/venv/bin/pip install \
        "git+https://github.com/python-escpos/python-escpos@f5ed42f2dcd1c19cfedb2e186a616701fa41ece8"
    build/external/python-escpos/venv/bin/python tools/external/python-escpos/capture.py

Then render what it wrote, which is what freezes the fixtures:

    node tools/external/python-escpos/capture.js

Nothing here runs during npm test.
"""

import json
import pathlib
import runpy
import sys
import urllib.request

import escpos.printer
from escpos.version import version as escpos_version

# The commit of the repository the examples are taken from

COMMIT = "f5ed42f2dcd1c19cfedb2e186a616701fa41ece8"

SOURCE = "https://github.com/python-escpos/python-escpos"

RAW = "https://raw.githubusercontent.com/python-escpos/python-escpos"

# Where this script lives, and the two directories it uses

ROOT = pathlib.Path(__file__).resolve().parents[3]

FIXTURES = ROOT / "test" / "fixtures" / "external" / "python-escpos"

# The examples are downloaded here, outside the repository, and not committed

DOWNLOADS = ROOT / "build" / "external" / "python-escpos"

# The virtual environment this script runs in, as the provenance records it: the
# interpreter of `command`, and the two steps of `setup` that build it

VENV = "build/external/python-escpos/venv"

PYTHON = f"{VENV}/bin/python"

SETUP = (
    f"python3 -m venv {VENV} && "
    f'{VENV}/bin/pip install "git+{SOURCE}@{COMMIT}"'
)

# The examples that run against a Dummy printer without hardware or network.
# `argv` is what the example is given on the command line, `assets` are the
# files it reads next to itself, and `features` is what it exercises, which the
# provenance records.

EXAMPLES = {
    "barcodes": {
        "file": "examples/barcodes.py",
        "features": ["hardware barcodes", "software barcodes", "GS k Code 39"],
    },
    "block_text": {
        "file": "examples/block_text.py",
        "features": ["ESC ! print modes", "custom size", "underline", "block text", "partial cut"],
    },
    "font_variations": {
        "file": "examples/font_variations.py",
        "features": ["ESC ! print modes", "GS ! sizes", "bold", "underline", "alignment", "partial cut"],
    },
    "qr_code": {
        "file": "examples/qr_code.py",
        "argv": ["https://github.com/python-escpos/python-escpos"],
        "features": ["GS ( k native QR code", "centring"],
    },
    "receipt": {
        "file": "examples/receipt.py",
        "assets": ["examples/graphics/receipt/hocus-pocus.gif", "examples/graphics/receipt/creature5.gif"],
        "features": ["images", "line spacing", "GS ! sizes", "bold", "underline",
                     "alignment", "software columns", "partial cut"],
    },
    "software_barcode": {
        "file": "examples/software_barcode.py",
        "features": ["software barcodes as bitImageRaster", "software barcodes as the default image implementation"],
    },
    "software_columns": {
        "file": "examples/software_columns.py",
        "features": ["software columns", "font A"],
    },
}

# The examples of the directory that are not captured, with the reason

SKIPPED = {
    "weather": "fetches a forecast over the network",
    "docker-flask": "a web service, not a receipt",
    "codepage_tables": "passes strings to _raw(), which Dummy.output cannot join in this version",
}

# The printer classes that are replaced with a Dummy while an example runs

PRINTERS = ["Dummy", "Usb", "Network", "Serial", "File", "CupsPrinter", "LP", "Win32Raw"]

# The printer a Usb example is given instead of its device, per example. An
# example that constructs its own Dummy keeps the profile it asks for.

USB_PROFILES = {
    "barcodes": "TM-T88II",
    "qr_code": "POS-5890",
}


def download(path):
    """Fetch a file of the repository at the pinned commit, cached on disk."""
    target = DOWNLOADS / path

    if not target.exists():
        target.parent.mkdir(parents=True, exist_ok=True)

        with urllib.request.urlopen(f"{RAW}/{COMMIT}/{path}") as response:
            target.write_bytes(response.read())

    return target


def capture(name, example):
    """Run one example against a Dummy printer and return what it wrote."""
    script = download(example["file"])

    for asset in example.get("assets", []):
        download(asset)

    created = []

    class Captured(escpos.printer.Dummy):
        """A Dummy that registers itself, so that the bytes can be collected."""

        def __init__(self, *args, **kwargs):
            profile = kwargs.get("profile") or USB_PROFILES.get(name)
            super().__init__(profile=profile)
            created.append(self)

    # Every printer class an example may construct becomes this Dummy, so that
    # the example runs unchanged and without a device. The originals are put
    # back afterwards, so that the next example subclasses the library's own
    # Dummy again instead of the one of the example before it

    patched = [name for name in PRINTERS if hasattr(escpos.printer, name)]
    originals = {name: getattr(escpos.printer, name) for name in patched}

    argv = sys.argv
    sys.argv = [str(script)] + example.get("argv", [])

    try:
        for attribute in patched:
            setattr(escpos.printer, attribute, Captured)

        runpy.run_path(str(script), run_name="__main__")
    finally:
        for attribute, original in originals.items():
            setattr(escpos.printer, attribute, original)

        sys.argv = argv

    if not created:
        raise RuntimeError(f"{name} created no printer")

    output = b"".join(printer.output for printer in created)

    profile = created[0].profile

    return output, profile


def main():
    """Capture every example this script was asked for."""
    only = sys.argv[1:]
    names = [name for name in EXAMPLES if not only or name in only]

    if not names:
        raise SystemExit(f"No such example, one of {', '.join(EXAMPLES)}")

    FIXTURES.mkdir(parents=True, exist_ok=True)

    index = {}

    print(f"python-escpos {escpos_version}")

    for name in names:
        example = EXAMPLES[name]
        output, profile = capture(name, example)

        (FIXTURES / f"{name}.bin").write_bytes(output)

        media = profile.profile_data.get("media", {}).get("width", {}).get("pixels")
        columns = profile.get_columns("a")

        index[name] = {
            "file": example["file"],
            "commit": COMMIT,
            "source": SOURCE,
            "version": escpos_version,
            "profile": profile.profile_data.get("name") or "default",
            "columns": columns,
            "media": media if isinstance(media, int) else None,
            "argv": example.get("argv", []),
            "features": example["features"],
            "usb": name in USB_PROFILES,
            "setup": SETUP,
            "python": PYTHON,
        }

        print(f"  {name:<24} {len(output):>7} bytes  profile {index[name]['profile']}")

    # What the Node side needs to write the provenance, next to the streams. The
    # entries are merged into what is there, so that capturing one example again
    # does not throw the index of the others away

    path = DOWNLOADS / "index.json"
    merged = json.loads(path.read_text()) if path.exists() else {}

    merged.update(index)

    path.write_text(json.dumps(merged, indent=2) + "\n")

    print(f"\nskipped: {', '.join(f'{name} ({reason})' for name, reason in SKIPPED.items())}")


if __name__ == "__main__":
    main()

File: tools/external/python-escpos/capture.js
import fs from 'node:fs';
import path from 'node:path';
import {fileURLToPath} from 'node:url';

import {write, report, today, directory, render} from '../shared.js';

/*
    python-escpos, MIT, https://github.com/python-escpos/python-escpos

    The widest command usage of the libraries of section 16: print modes, tabs
    and absolute positions, its three image implementations, codepage switching
    through its magic encode, barcodes and QR codes.

    The byte streams are captured by the Python script next to this one, which
    runs the examples of the repository against python-escpos' Dummy printer in
    a throwaway virtual environment and writes the .bin files and an index of
    what every example assumed:

        python3 -m venv build/external/python-escpos/venv
        build/external/python-escpos/venv/bin/pip install \
            "git+https://github.com/python-escpos/python-escpos@<commit>"
        build/external/python-escpos/venv/bin/python tools/external/python-escpos/capture.py

    This script renders what it wrote and freezes the fixtures:

        node tools/external/python-escpos/capture.js

    It reads the index the Python script left in build/external/python-escpos,
    so the two run in that order. Nothing here runs during npm test.
*/

const LIBRARY = 'python-escpos';

const root = path.join(path.dirname(fileURLToPath(import.meta.url)), '..', '..', '..');

/* Where the Python script left its index of the capture */

const INDEX = path.join(root, 'build', 'external', LIBRARY, 'index.json');

/*
    python-escpos writes its codepage commands as ESC t n out of the codePages
    table of the profile, which is the Epson table: 0 is cp437, 2 cp850,
    16 windows1252, 19 cp858, and so on. That is the `epson` mapping.
*/

const MAPPING = 'epson';

/* The width of a character cell of font A, which is what turns the columns of
   a profile into a print width when the profile does not give one */

const CHARACTER_WIDTH = 12;

/* What the unknown items of these fixtures are. python-escpos sends one command
   this renderer does not implement, and every fixture that has unknown items has
   only that one */

const UNKNOWN = 'The unknown items are GS b, the smoothing mode, which set_with_default() sends with every ' +
  'style change and which changes nothing on paper.';

/**
 * The print width a capture is rendered at, in dots.
 *
 * The profiles of python-escpos carry the width of the paper in dots and the
 * number of columns of font A separately, and the two do not always agree at
 * the twelve dot cell this renderer draws font A in: a TM-U220 is 400 dots wide
 * and 42 columns, because its font is narrower than a thermal one. The paper is
 * what the printer has, so the paper wins, rounded down to whole bytes.
 *
 * @param  {object}   entry   One entry of the index the Python script wrote
 * @return {number}           The width in dots
 */
function widthOf(entry) {
  const dots = entry.media || entry.columns * CHARACTER_WIDTH;

  return dots - (dots % 8);
}

/**
 * What the notes of a fixture say: the profile the example assumed, the
 * features it exercises and how it was run
 *
 * @param  {string}   name      Name of the fixture
 * @param  {object}   entry     One entry of the index
 * @param  {number}   unknown   Number of unknown items the render produced
 * @return {string}             The notes
 */
function notes(name, entry, unknown) {
  const width = widthOf(entry);
  const columns = Math.floor(width / CHARACTER_WIDTH);

  return [
    `Profile ${entry.profile}, ${entry.columns} columns of font A and ${entry.media || 'no'} dots of paper.`,
    entry.media && columns !== entry.columns ?
      `The renderer draws font A in twelve dots, so the paper holds ${columns} columns instead of ` +
      `the ${entry.columns} the profile reports; python-escpos lays its software columns out for ${entry.columns}.` :
      null,
    `Exercises ${entry.features.join(', ')}.`,
    entry.usb ?
      'The example constructs a Usb printer, which the capture replaced with a Dummy of the same profile.' :
      null,
    'Reviewed as a PNG against what the example prints.',
    unknown ? UNKNOWN : null,
  ].filter((note) => note).join(' ');
}

/**
 * Freeze the fixtures the Python script captured
 *
 * @param  {string[]}   only   Names of the fixtures, empty for all of them
 */
function main(only) {
  if (!fs.existsSync(INDEX)) {
    throw new Error(`No capture to render, run tools/external/${LIBRARY}/capture.py first`);
  }

  const index = JSON.parse(fs.readFileSync(INDEX, 'utf8'));
  const names = Object.keys(index).filter((name) => only.length === 0 || only.includes(name));

  if (!names.length) {
    throw new Error(`No such fixture, one of ${Object.keys(index).join(', ')}`);
  }

  console.log(`${LIBRARY} ${index[names[0]].version}`);

  for (const name of names) {
    const entry = index[name];
    const bytes = new Uint8Array(fs.readFileSync(path.join(directory, LIBRARY, `${name}.bin`)));

    const argv = entry.argv.length ? ` ${entry.argv.map((value) => JSON.stringify(value)).join(' ')}` : '';

    /* The notes say what the unknown items are, so the fixtures without any do
       not claim to have them: rendering once up front answers that */

    const provenance = {
      language: 'esc-pos',
      width: widthOf(entry),
      codepageMapping: MAPPING,
    };

    const unknown = render(bytes, provenance).filter((item) => item.type === 'unknown').length;

    const result = write(LIBRARY, name, bytes, {
      source: entry.source,
      file: entry.file,
      commit: entry.commit,
      licence: 'MIT',
      setup: entry.setup,
      command: `${entry.python} tools/external/${LIBRARY}/capture.py ${name}${argv} && ` +
        `node tools/external/${LIBRARY}/capture.js ${name}`,
      version: entry.version,
      language: 'esc-pos',
      columns: Math.floor(widthOf(entry) / CHARACTER_WIDTH),
      width: widthOf(entry),
      codepageMapping: MAPPING,
      captured: today(),
      notes: notes(name, entry, unknown),
    });

    console.log(report(name, result));
  }
}

if (process.argv[1] === new URL(import.meta.url).pathname) {
  main(process.argv.slice(2));
}
